File: src/types/message.ts
import type { UserProfile } from './user';

export interface Message {
  id: string;
  chatId: string;
  senderId: string;
  content: string;
  messageType: 'text' | 'image' | 'video' | 'audio' | 'file' | 'location' | 'contact' | 'sticker' | 'system';
  replyToId?: string;
  forwardedFromId?: string;
  metadata?: Record<string, any>;
  isEdited: boolean;
  isDeleted: boolean;
  deletedAt?: Date;
  createdAt: Date;
  updatedAt: Date;
}

export interface MessageWithSender extends Message {
  sender: UserProfile;
  replyTo?: Message;
  attachments: Attachment[];
  reactions: MessageReaction[];
  readBy: MessageRead[];
}

export interface Attachment {
  id: string;
  messageId: string;
  fileName: string;
  fileSize: number;
  mimeType: string;
  url: string;
  thumbnailUrl?: string;
  duration?: number;
  width?: number;
  height?: number;
  createdAt: Date;
}

export interface MessageRead {
  id: string;
  messageId: string;
  userId: string;
  readAt: Date;
}

export interface MessageReaction {
  id: string;
  messageId: string;
  userId: string;
  emoji: string;
  createdAt: Date;
}

export interface SendMessageData {
  chatId: string;
  content: string;
  messageType: Message['messageType'];
  replyToId?: string;
  attachments?: AttachmentInput[];
}

export interface AttachmentInput {
  fileName: string;
  fileSize: number;
  mimeType: string;
  url: string;
  thumbnailUrl?: string;
  duration?: number;
  width?: number;
  height?: number;
}

export interface TypingIndicator {
  chatId: string;
  userId: string;
  isTyping: boolean;
}

File: crystal_messenger/lib/models/message.dart
import 'package:freezed_annotation/freezed_annotation.dart';

part 'message.freezed.dart';
part 'message.g.dart';

enum MessageType { text, image, video, audio, file, location, contact, sticker, system }

@freezed
class Message with _$Message {
  const factory Message({
    required String id,
    @JsonKey(name: 'chat_id') required String chatId,
    @JsonKey(name: 'sender_id') required String senderId,
    String? content,
    @JsonKey(name: 'message_type') @Default(MessageType.text) MessageType messageType,
    @JsonKey(name: 'reply_to_id') String? replyToId,
    @JsonKey(name: 'forwarded_from_id') String? forwardedFromId,
    @Default(false) bool isDeleted,
    @JsonKey(name: 'created_at') required DateTime createdAt,
    @JsonKey(name: 'delivered_at') DateTime? deliveredAt,
    @JsonKey(name: 'read_at') DateTime? readAt,
    Map<String, dynamic>? metadata,
    @JsonKey(name: 'reply_to') Message? replyTo,
    List<Attachment>? attachments,
  }) = _Message;

  factory Message.fromJson(Map<String, dynamic> json) => _$MessageFromJson(json);
}

@freezed
class Attachment with _$Attachment {
  const factory Attachment({
    required String id,
    @JsonKey(name: 'message_id') required String messageId,
    @JsonKey(name: 'file_name') required String fileName,
    @JsonKey(name: 'file_size') required int fileSize,
    @JsonKey(name: 'mime_type') required String mimeType,
    required String url,
    @JsonKey(name: 'thumbnail_url') String? thumbnailUrl,
    int? duration,
    int? width,
    int? height,
  }) = _Attachment;

  factory Attachment.fromJson(Map<String, dynamic> json) =>
      _$AttachmentFromJson(json);
}
